sycl/doc/extensions/proposed/sycl_ext_oneapi_peer_access.asciidoc

```diff
@@ -0,0 +1,162 @@
+direct memory access between peer devices in a system.
+In particular, this allows one device to directly access USM Device
+allocations for a peer device in the same context.
+Peer to peer capabilities are useful as they can provide access to a peer
+device's memory inside a compute kernel and also optimized memory copies between
+peer devices.
+
+This extension adds the following new member functions to the device class, as described
+below.
+
+[source,c++]
+----
+namespace sycl   {
+namespace ext    {
+namespace oneapi {
+enum class peer_access {
+  access_supported,
+  access_enabled,
+  atomics_supported,
+};
+} // namespace oneapi
+} // namespace ext
+
+class device {
+public:
+  bool ext_oneapi_can_access_peer(const device &peer,
+                                  ext::oneapi::peer_access value =
+                                    ext::oneapi::peer_access::access_supported);
+  void ext_oneapi_enable_peer_access(const device &peer);
+  void ext_oneapi_disable_peer_access(const device &peer);
+};
+
+} // namespace sycl
+----
+
+The semantics of the new functions are:
+
+|===
+|Member Function |Description
+
+|bool ext_oneapi_can_access_peer(const device &peer,
+                                  ext::oneapi::peer_access value =
+                                  ext::oneapi::peer_access::access_supported)
+a|Queries the peer access status between this device and `peer` according to
+the query `value`:
+
+* `ext::oneapi::peer_access::access_supported`: Returns true only if it is
+possible for this device to enable peer access to USM device memory allocations
+located on the `peer` device.
+
+* `ext::oneapi::peer_access::access_enabled`: Returns true only if peer access is
+currently enabled from this device to the `peer` device.
+
+* `ext::oneapi::peer_access::atomics_supported`: When this query returns true,
+it indicates that this device may concurrently perform atomic operations on USM
+device memory allocations located on the `peer` device when peer access is enabled
+to that device. Atomics performed on a peer device's memory must have
+`memory_scope::system` scope.
+If the query returns false, attempting to perform atomic operations
+on `peer` memory will have undefined behavior.
+
+|void enable_peer_access(const device &peer)
+|Enables this device to access USM device allocations located on the peer
+device. This does not permit the peer device to access this device's memory.
+Once this access is enabled, SYCL kernel functions and the explicit memory
+functions may access USM device allocations on the peer device subject to the
+normal rules about context as described in the core SYCL specification.
+Throws an exception if access cannot be enabled or if access is already
+enabled.
+
+|void disable_peer_access(const device &peer)
+|Disables access to the peer device's memory from this device. Throws an
+exception if access cannot be disabled or if access is not enabled.
```

Comment:
Similar wording here:

> If peer access is already disabled, throws an `exception` with the `errc::invalid` error code.

I know the wording previous said that an exception is thrown if access cannot be disabled.  Would this be due to some system-level error that is outside of the application's control?  I think it is assumed that any SYCL API could fail for a reason like this with an implementation-specific error code.  If you feel that needs to be noted specifically for this API, the description could be like this (but I think it is unnecessary):

> If peer access is already disabled, throws an `exception` with the `errc::invalid` error code.  This function could also throw an `exception` with an implementation-defined error code if the underlying backend is unable to disable peer access.

Reply:
Good question. In my draft impl this happens in the Cuda backend if I call `Devs[0].ext_oneapi_disable_peer_access(Devs[1]);` without enabling the access prior to this call:

```
jakirk@cori04:~/dpcCode> ./res

PI CUDA ERROR:
	Value:           705
	Name:            CUDA_ERROR_PEER_ACCESS_NOT_ENABLED
	Description:     peer access has not been enabled
	Function:        cuda_piextDisablePeer
	Source Location: /global/u1/j/jakirk/llvm/sycl/plugins/cuda/pi_cuda.cpp:5638

terminate called after throwing an instance of 'sycl::_V1::runtime_error'
  what():  Native API failed. Native API returns: -999 (Unknown PI error) -999 (Unknown PI error)

```